### PR TITLE
support A3 projects with npm workspaces packages

### PR DESCRIPTION
**Pull request by haroun**
## Summary

Add support for direct workspace dependencies

## What are the specific steps to test this change?

You run the test `test/workspaces-project.js` or

1. Start from `a3-boilerplate`
2. Run `npm init -w workspace-a -y` to create `workspace-a`
3. Install `@apostrophecms-pro/scheduled-publishing` in `workspace-a` with `npm install -w workspace-a --save @apostrophecms-pro/scheduled-publishing`
4. In the `a3-boilerplate` project, enable the module `@apostrophecms-pro/scheduled-publishing`. It should not be present in the `a3-boilerplate` `package.json`. It should pass.
5. Switch to `apostrophe` `main` branch, the same thing should fail with `The configured bundle @apostrophecms-pro/scheduled-publishing was not found in npm`

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [x] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [ ] The changelog is updated
- [ ] Related documentation has been updated
- [ ] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: lib/moog-require.js [hunk 1]
```diff
@@ -7,6 +7,30 @@ const importFresh = require('import-fresh');
 const resolveFrom = require('resolve-from');
 const regExpQuote = require('regexp-quote');
 
+const getDependencies = ({
```
Comment (boutell) [inline lib/moog-require.js:10]: Use `function` and place these at the end, don't front-load the file with implementation detail code that makes it harder to read the overall intent.

File: lib/moog-require.js [hunk 2]
```diff
@@ -7,6 +7,30 @@ const importFresh = require('import-fresh');
...
+    ...Object.keys(devDependencies || {})
+  ];
+};
+
+const getWorkspacesDependencies = ({ workspaces = [] } = {}) => {
+  if (workspaces.length === 0) {
+    return [];
+  }
+
+  const workspacePackages = glob.sync(`{${workspaces.join(',')}}/package.json`, { follow: true });
```
Comment (boutell) [inline lib/moog-require.js:25]: Do not make assumptions about the current working directory. Note how the code that calls `getWorkspacesDependencies` figures out what folders to look at on each iteration and pass `folder` into `getWorkspacesDependencies` so you can concatenate it with the workspace name.

**Review by boutell — CHANGES_REQUESTED** [review]
General strategy makes sense, see notes. Thanks

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## UNRELEASED
+
+### Adds
+
+* Add support for npm workspaces dependencies. A workspace dependencies can now be used as an Apostrophe module even if it is not a direct dependency of the Apostrophe project. Only direct workspaces dependencies of the Apostrophe project are supported. e.g. I have an Apostrophe project called `website`. `website` is set with 2 workspaces, `workspace-a` & `workspace-b`. `workspace-a` `package.json` contains a module named `blog` as a dependency. `website` can reference `blog` as enabled in the `modules` configuration.
```
Comment (boutell) [inline CHANGELOG.md:7]: A workspaces dependencies -> A workspaces dependency

"For instance, I have..."

"With 2 workspaces" -> "with two npm workspaces"

Link "npm workspaces" to relevant npm documentation on the first mention

(And sync up with Bob for docs)

File: lib/moog-require.js [hunk 4]
```diff
@@ -231,7 +234,35 @@ module.exports = function(options) {
...
+    ];
+  }
+
+  function getWorkspacesDependencies(folder) {
+    return ({ workspaces = [] } = {}) => {
+      if (workspaces.length === 0) {
+        return [];
+      }
+
+      const pattern = workspaces.length === 1 ? workspaces[0] : `{${workspaces.join(',')}}`;
```
Comment (boutell) [inline lib/moog-require.js:255]: Pretty sure the join call is always sufficient here

File: lib/moog-require.js [hunk 5]
```diff
@@ -231,7 +234,35 @@ module.exports = function(options) {
...
+      const packagePath = path.resolve(folder, pattern, 'package.json');
+      const workspacePackages = glob.sync(packagePath, { follow: true });
```
Comment (boutell) [inline lib/moog-require.js:257]: niiice

Comment (boutell) [reply lib/moog-require.js:257]: Very slick use of advanced glob features

**Review by boutell — CHANGES_REQUESTED** [review]

File: lib/moog-require.js (hunk 4 above)
Comment (haroun) [reply lib/moog-require.js:256]: I've added a comment, glob expects at least two entries when using curly braces. Silly bug!

File: test/workspaces-project/modules/@apostrophecms/log/index.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+const createLogger = () => {
```
Comment (boutell) [inline test/workspaces-project/modules/@apostrophecms/log/index.js:1]: good technique.

**Review by boutell — APPROVED** [review]

**Review by boutell — APPROVED** [review]

**Pull request merged**